Ranks: datums and manifests

DeeUnderscore opened the pull request:
This PR changes the way ranks are handled in order to make them more flexible. It also uses these changes to display ranks in the crew manifest. 
## Rank datums
- Ranks and branches are now in datums. 
- A rank datum currently holds the full name, and the abbreviated name of he rank. A branch datum has the same fields. It could possibly also hold something like rank insignia in the future.
- Each branch specifies a list of ranks which apply to that branch, as well as a list of spawn ranks, a subset of the above which players can choose for their characters before spawning.
- Each map datum specifies a list of branches and spawn branches, analogous to the above.
- With this arrangement you need at least one branch in order to have ranks in a map. This simplifies things from the code side and is not a problem with the Torch, but requires the use of one branch even if a map were to be branchless.
## Use of rank datums and ranks
- There are now extra checks to disallow latejoining with incompatible ranks or branches. Job preferences are also pruned to ensure incompatible jobs are not in preferences. 
- Ranks and branches for people are now held in the datacore. The employee records terminal currently does not have a way to change those, but it should be easy to add in the nearest future. 
- Crew manifests now include people's ranks. They appear as prefixes to the name, and use `<abbr>`s with tooltips containing the full name of the rank, and the full name of the branch. Civilians do not get a prefix to their name (their ranks' short names are `null`).
  ![manifest-ranks](https://cloud.githubusercontent.com/assets/156239/19419110/d314c7ea-93d0-11e6-91dc-5ad19b1fabf3.png)
## Torch ranks
- I added abbreviated name to every rank specified in Torch map files. The Fleet abbreviations are adapted from US Navy abbreviations, while Marine Corps abbreviations are those used by the USMC. This means that Fleet abbreviations are all caps, while Marine ones are not.

DeeUnderscore committed:
Datumize ranks

DeeUnderscore committed:
Extra checks for branch/rank job compatibility

Latejoin job list will now omit jobs incompatible with current rank.
Changing ranks or branches will now remove incompatible jobs from the
player's job preferences.

DeeUnderscore committed:
Add mil rank and branch to datacore

DeeUnderscore committed:
Add rank to PDA manifest

Abbreviated ranks come before the name, and <abbr> is used for full rank
name and branch.

DeeUnderscore committed:
Add ranks to non-PDA manifest

DeeUnderscore committed:
Pull out rank handling into a global object

DeeUnderscore committed:
Add mil_branches to the debug controller verb

PsiOmegaDelta commented on code/modules/client/preferences.dm:
```diff
@@ -387,6 +387,9 @@ datum/preferences
 	character.religion = religion
+	
+	character.char_branch = char_branch
+	character.char_rank = char_rank
```
Here seems like a good opportunity to switch from the client having to deal with strings to use the proper datums instead.

PsiOmegaDelta replied:
Here seems like a good opportunity to switch from the client having to deal with strings to use the proper datums instead.

Finalsong commented:
This is great.
Also: An XO with the surname Tigh, you say? I smell BSG

Raptor1628 commented:
Approve the appropriate use of Abbreviations. (And also alliteration) The moment this is merged I can get back to setting up Torch jobs.

DeeUnderscore committed:
Store player's branch and rank as reference

Datacore and human mobs store references to rank and branch instead of
storing string keys to them.

PsiOmegaDelta commented on code/datums/datacore.dm:
```diff
@@ -230,6 +241,8 @@
 		L.fields["religion"]	= H.religion
 		L.fields["image"]		= getFlatIcon(H)	//This is god-awful
+		L.fields["mil_branch"]   = H.char_branch
+		L.fields["mil_rank"]    = H.char_rank
```
Now that these are datums shouldn't this be H.char_rank.name or similar?
Don't forget that the human define definitions need to have their full paths to access the vars.

DeeUnderscore replied:
I figured I might as well store references to the datums in datacore instead of string keys. Either way, there has to be some casting involved along the way since either way they are in lists. Might as well use strings here, though.
